File: DONGUK/2023/n+1_카드_게임.py
def getNumToAdd(hands, target):
    for hand in hands:
        if target - hand in hands:
            return hand
    return -1

def solution(coin, cards):
    answer = 0
    
    hands = set()
    trash = set()
    n = len(cards)
    target = n+1

    for _ in range(n//3):
        card = cards.pop(0)
        hands.add(card)
    
    while True:
        answer += 1
        # print('hands ', hands, ' trash ', trash)
        if not cards:
            break
            
        card1 = cards.pop(0)
        card2 = cards.pop(0)
        trash.add(card1)
        trash.add(card2)

        # 코인 0개 사용 -> 손에 있는 패 털기
        num = getNumToAdd(hands, target)
        if num > 0:
            hands.remove(num)
            hands.remove(target - num)
            continue
                
        # 코인 1개 사용 -> 손에 있는 카드 1개, 휴지통에서 1개
        flag = False
        for card in hands:
            if target - card in trash and coin >= 1:
                trash.remove(target - card)
                hands.remove(card)
                coin -= 1
                flag = True
                break
        
        if flag:
            continue
            
        if coin >= 2:
            card = getNumToAdd(trash, target)
            if card != -1:
                trash.remove(card)
                trash.remove(target - card)
                coin -= 2
                continue

        break
            
    return answer

File: JINSEOP/프로그래머스_혼자_풀기/n + 1 카드 게임.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <iostream>
using namespace std;

bool card_check[1001];
bool validate_hand(vector<int> &hands, int goal) {
	for (int card : hands) {
		if (find(hands.begin(), hands.end(), goal - card) != hands.end()) {
			hands.erase(remove(hands.begin(), hands.end(), card), hands.end());
			hands.erase(remove(hands.begin(), hands.end(), goal - card), hands.end());
			card_check[card] = 0;
			card_check[goal - card] = 0;

			return true;
		}
	}

	return false;
}

bool validate_draw_one(vector<int>& hands, int goal) {
	for (int card : hands) {
		if (card_check[goal - card]) {
			hands.erase(remove(hands.begin(), hands.end(), card), hands.end());
			card_check[card] = 0;
			card_check[goal - card] = 0;

			return true;
		}
	}

	return false;
}

bool validate_draw_two(int goal) {
	for (int k = 1; k < goal; k++) {
		if (card_check[k] && card_check[goal - k]) {
			card_check[k] = 0;
			card_check[goal - k] = 0;

			return true;
		}
	}

	return false;
}

int solution(int coin, vector<int> cards) {
	int answer = 1, goal = cards.size() + 1;
	vector<int> hands;

	for (int i = 0; i < cards.size() / 3; i++) {
		hands.push_back(cards[i]);
		// n/3 장의 카드를 첫 패로 뽑는다.
		card_check[cards[i]] = true;
	}

	for (int i = cards.size() / 3; i < cards.size(); i += 2) {
		// 남은 카드 두 장씩 훑는다
		card_check[cards[i]] = true;
		card_check[cards[i + 1]] = true;

		// 1. 손에 있는 두 패로 n + 1을 만들 수 있는 경우
		if (hands.size() >= 2 && validate_hand(hands, goal))
			answer++;
		// 2. 손에 있는 패 1장, 뽑은 카드 1장으로 n + 1을 만들 수 있는 경우
		else if (hands.size() >= 1 && coin >= 1 && validate_draw_one(hands, goal)) {
			answer++;
			coin--;
		}
		// 3. 손에 있는 패를 모두 써야 만들 수 있는 경우
		else if (coin >= 2 && validate_draw_two(goal)) {
			answer++;
			coin -= 2;
		}
		else
			break;
	}

	return answer;
}
